x-pack/plugin/sql/sql-client/src/main/java/org/elasticsearch/xpack/sql/client/ClientVersion.java

```diff
@@ -71,23 +72,36 @@
         CURRENT = extractVersion(url);
     }
 
+    // There are three main types of provided URLs:
+    // (1) a file URL: file:<path><FS separator><driver name>.jar
+    // (2) jar file URL pointing to a JAR file: jar:<sub-url><separator><driver name>.jar!/
+    // (3) jar file URL pointing to a JAR file entry (likely a fat JAR, but other types are possible): jar:<sub-url>!/driver name>.jar!/
     static SqlVersion extractVersion(URL url) {
+        Manifest manifest = null;
         String urlStr = url.toString();
         if (urlStr.endsWith(".jar") || urlStr.endsWith(".jar!/")) {
             try {
                 URLConnection conn = url.openConnection();
-                conn.setUseCaches(false);
-
-                try (JarInputStream jar = new JarInputStream(conn.getInputStream())) {
-                    Manifest manifest = jar.getManifest();
-                    String version = manifest.getMainAttributes().getValue("X-Compile-Elasticsearch-Version");
-                    return SqlVersion.fromString(version);
-                }
+                do {
+                    // For a jar protocol, the implementing java.base/sun.net.www.protocol.jar.JarUrlConnection#getInputStream() will only
+                    // return a stream (vs. throw an IOException) if the JAR file URL points to a JAR file entry and not a JAR file.
+                    if (url.getProtocol().equals("jar")) {
+                        JarURLConnection jarConn = (JarURLConnection) conn;
+                        if (jarConn.getEntryName() == null) { // the URL points to a JAR file
+                            manifest = jarConn.getManifest(); // in case of a fat JAR, this would return the outermost JAR's manifest
+                            break;
+                        }
+                    }
+                    try (JarInputStream jar = new JarInputStream(conn.getInputStream())) {
+                        manifest = jar.getManifest();
+                    }
+                } while (false);
```

Comment:
`do { } while (false)` or use a function with the same break conditions (but using `return` to "jump" at the end)?

Reply:
This section needs improving.
the `do {} while (false)` is unnecessary. The break can also be removed by checking if the manifest is still null before the try block:

```
if (url.getProtocol()...) {
}

if (manifest == null) {
 try (JarInputStream...)
}
```

Further more, if the entry is not-null what would be the returned manifest?
```
if (url.getProtocol().equals("jar")) {
   JarURLConnection jarConn = (JarURLConnection) conn;
   manifest = jarConn.getManifest(); <-- ? what would this return?

```

Reply:
> do { } while (false)

I thought this might raise some eyebrows. :-)

> or use a function with the same break conditions (but using return to "jump" at the end)?

I've turned the frame loop into a function. Hope this is what you were suggesting.

> the do {} while (false) is unnecessary.

Generally this avoids an unnecessary fetch&compare. 
But in this case if the manifest is still `null` (shouldn't, but if), `conn.getInputStream()` would raise an exception (vs. just returning the "null version").

> if the entry is not-null what would be the returned manifest?

If the entry is not null, it'd simply return the [outermost] JAR's manifest (since the entry can be any class, i.e. non a JAR). Let me know if I still didn't get your question, though.